tests/tools/s3/__init__.py

```diff
@@ -0,0 +1,28 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+import json
+import os
+import pathlib
+
+from .utils import S3ResourceFetcher
+
+
+def download_resource(resource_type, resource_name, resource_tags="{}", path=None):
```

Comment:
Since this is a simple utility function, it does not matter much, but from an API point of view, I  would have `path` as always a required option to this function.; For instance it's really the responsibility of the caller to provide the path to download to (and whether to use a default or a user provided path is what a caller should decide). 

Also, `download_resource` should ideally return a single resource. So that the logic of suitably sorting downloaded resources (and returning the first) remains well within the class itself and the caller just gets one resource. Rather than caller taking out the first resource explicitly. 

And if indeed if multiple are are required to be downloaded (a less common use case), this and the underlying `download` should take `many=False` optional argument. So the caller should specify if they are interested in `many` resources.

Reply:
> Since this is a simple utility function, it does not matter much, but from an API point of view, I would have `path` as always a required option to this function.; For instance it's really the responsibility of the caller to provide the path to download to (and whether to use a default or a user provided path is what a caller should decide).
> 

I am adept of as few parameters as possible, and providing good defaults. In Python this is even easy to achieve. In the normal use case, you should not care where the resources are downloaded because the script already tells you where they are placed, and the location should be good enough for the general scenario (i.e. they're downloaded in `resources` which is the same directory used for the scripts that build said resources). I think only special scenarios require a different location (like for example you want to keep a cache of resources somewhere not in the repository because the repository gets cleaned up after each run). 
 
> Also, `download_resource` should ideally return a single resource. So that the logic of suitably sorting downloaded resources (and returning the first) remains well within the class itself and the caller just gets one resource. Rather than caller taking out the first resource explicitly.
> 

This is not really what I had in mind for this, but then it is also not very well documented. This function is suppose to return all resources that have the specified tags and types (the name should be removed from the required list). That is because I am expecting the test code to evolve such that we parameterize the tests easier than what we do now. For example, let's say that we have support for 2 distros (Ubuntu & Fedora). In the tests I would like to check that all of the available distros are booting under a specified time. Then what I would need to do is the following:
- download_resource(type="kernel", tags="{\"halt-after-boot\": False, \"distro\" = \"*\"} -> ["~/resources/kernel/elf/ubuntu", "~/resources/kernel/elf/fedora", "~/resources/kernel/bzimage/fedora", "~/resources/kernel/elf/ubuntu"]
- for each returned kernel, I want to query all disk resources, that "work-with-kernel":
```
distro_kernels = ..
for kernel in distro_kernel:
     disk = download_resource(type="disk", tags="{\"works-with-kernel\":{}}".format(kernel))
     # add the pair (disk, kernel) to a the list used for parametrizing the test.
```
> And if indeed if multiple are are required to be downloaded (a less common use case), this and the underlying `download` should take `many=False` optional argument. So the caller should specify if they are interested in `many` resources.

I think it is a good idea to have a way of only downloading one. How about doing this reversed though, and have a parameter to just get the first one that matches all tags?

Reply:
> > Since this is a simple utility function, it does not matter much, but from an API point of view, I would have `path` as always a required option to this function.; For instance it's really the responsibility of the caller to provide the path to download to (and whether to use a default or a user provided path is what a caller should decide).
> 
> I am adept of as few parameters as possible, and providing good defaults. In Python this is even easy to achieve. In the normal use case, you should not care where the resources are downloaded because the script already tells you where they are placed, and the location should be good enough for the general scenario (i.e. they're downloaded in `resources` which is the same directory used for the scripts that build said resources). I think only special scenarios require a different location (like for example you want to keep a cache of resources somewhere not in the repository because the repository gets cleaned up after each run).

Yes, I agree with keeping sensible defaults, so may be it's a default at the outer API level, but the class`s download function (the last one in the call chain) should have it mandatory and download location really need not be a state of the class. In this case it may not matter much because it's the only use of the class 'to download for the tests' or cache.  

The class should simply be a wrapper over the underlying `boto` object corresponding to the `bucket` and `manifest`.  (and can be reused to upload as well, now when I think of it!) 

I also agree, this is more a matter of opinion. :-) 

> 
> > Also, `download_resource` should ideally return a single resource. So that the logic of suitably sorting downloaded resources (and returning the first) remains well within the class itself and the caller just gets one resource. Rather than caller taking out the first resource explicitly.
> 
> This is not really what I had in mind for this, but then it is also not very well documented. This function is suppose to return all resources that have the specified tags and types (the name should be removed from the required list). That is because I am expecting the test code to evolve such that we parameterize the tests easier than what we do now. For example, let's say that we have support for 2 distros (Ubuntu & Fedora). In the tests I would like to check that all of the available distros are booting under a specified time. Then what I would need to do is the following:
> 
>     * download_resource(type="kernel", tags="{"halt-after-boot": False, "distro" = "*"} -> ["~/resources/kernel/elf/ubuntu", "~/resources/kernel/elf/fedora", "~/resources/kernel/bzimage/fedora", "~/resources/kernel/elf/ubuntu"]
> 
>     * for each returned kernel, I want to query all disk resources, that "work-with-kernel":
> 
> 
> ```
> distro_kernels = ..
> for kernel in distro_kernel:
>      disk = download_resource(type="disk", tags="{\"works-with-kernel\":{}}".format(kernel))
>      # add the pair (disk, kernel) to a the list used for parametrizing the test.
> ```
> 

Got it - then may be the function needs some documentation. :-)  and be called `download_resources`

> > And if indeed if multiple are are required to be downloaded (a less common use case), this and the underlying `download` should take `many=False` optional argument. So the caller should specify if they are interested in `many` resources.
> 
> I think it is a good idea to have a way of only downloading one. How about doing this reversed though, and have a parameter to just get the first one that matches all tags?

Yeah then instead of `many=False` (default `first=False`) and caller specifies `first=True` if they want only one.